### DISCUSSION TITLE
standardizing environment variables so that tools can work together

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#613] [Dev Discussions]
Currently every development/packaging tool for Python seems to have its own way of doing everything. PyPA is trying to define a tool independent way of defining which tools to use for the distribution lifecycle (build, package, publish). What I would like to see is that tools start standardizing on some of the configuration options that can be represented in environment variable, and eventually in pyproject.toml. 

For example:
1. VIRTUAL_ENV  - this is specified by virtualenv as the currently activated venv - do the other venv related tools do this?
2. VIRTUAL_ENV_HOME - the directory where venvs should be created. Currently tools seem to have their own default locations all over the place. They also have their own ways of setting it.
3. PROJECT_HOME - the directory where projects should be created and found.

Virtualenvwrapper uses the above three, although `VIRTUAL_ENV_HOME` is `WORK_ON` home and they are happy to change it.
In addition there are some other variables that could make tools work more seamlessly together.
USER
USER_EMAIL
etc. 

Many of these are needed to build and publish to PyPI. They should be in the pyproject.toml. But there should be a pyproject.toml for the user  that provides default values. Instead of looking in git for information, the variables should be available for git to use if it is not specifically configured.

I hope this is enough to start a conversion and get your thoughts. 

I would like to have putup start using these now, and be one of the first tools to standardize on them. 
`putup myproject` should use PROJECT_HOME if myproject has no directory part. Less typing, and the project is created in the correct place regardless of the current directory. 
`VIRTUAL_ENV` is not used to determine if a venv was active. (I which I could remember exactly where). I think it was in some tests/system code.

**Comment by FlorianWilhelm** [comment]
I am not sure if I really get this. The e-mail address is in `setup.cfg` right now and will be in `pyproject.toml` in the future. At project creation time it makes sense to take this from git as most developers of set this in their git global config. We could fall back to the environment variables maybe but then we would could still not create a meaningful initial commit as the user has to set up git before running `putup`.

Not sure I understand the VIRTUAL_ENV, ... proposal. It would be cool if all virtualenv tools would adhere to the same standards but this is independent of PyScaffold or maybe only useful if `--venv` is used.  

About PROJECT_HOME, we could discuss this although I am not sure if people would use it. You say `putup my_proj` and the project appears under `$PROJECT_HOME/my_proj`. The next step would be any way to cd into it, right? So is this useful?

**Reply by julie777** [reply]
> I am not sure if I really get this. The e-mail address is in `setup.cfg` right now and will be in `pyproject.toml` in the future. At project creation time it makes sense to take this from git as most developers of set this in their git global config. We could fall back to the environment variables maybe but then we would could still not create a meaningful initial commit as the user has to set up git before running `putup`.

We already have some fallback code. I think I created an issue because part of it isn't working right now.

As for our storing name and email, again I was being too thorough and describing a result that already partially exists.
> 
> Not sure I understand the VIRTUAL_ENV, ... proposal. It would be cool if all virtualenv tools would adhere to the same standards but this is independent of PyScaffold or maybe only useful if `--venv` is used.
> 

True, but if we are invoking virtualenv we could use it, just as virtualenvwrapper does with `mkvirtualenv`

> About PROJECT_HOME, we could discuss this although I am not sure if people would use it. You say `putup my_proj` and the project appears under `$PROJECT_HOME/my_proj`. The next step would be any way to cd into it, right? So is this useful?

If PROJECT_HOME was defined it just means I don't have to type the path to create a project. 

Regarding putup, if putup fully prepared the project for development, creating a virtualenv, running pip install -e, etc. Then my workflow would be
```
putup myproject
workon myproject
```

`workon` is part of virtualenv wrapper and I have been using it forever. It has auto-completion and it activates the venv and then changes directory to the project directory. I just login and type `workon pys<tab><enter>` and my venv for pysk is active and my current directory is the project for pyscaffold. You might be supersized how habit forming this is.
